File: programmers/level2/go/이진_변환_반복하기.go
package main

import (
	"strconv"
	"strings"
)

var cur string

func removeZeros() int {
	cnt := 0
	var sb strings.Builder

	arr := []rune(cur)
	for _, c := range arr {
		if c == '0' {
			cnt++
		} else {
			sb.WriteRune(c)
		}
	}

	cur = sb.String()
	return cnt
}

func binaryChange() int {
	len := len(cur)
	var arr []string

	for len > 0 {
		arr = append([]string{strconv.Itoa(len % 2)}, arr...)
		len /= 2
	}

	cur = strings.Join(arr, "")

	return 1
}

func 이진_변환_반복하기(s string) []int {
	cur = s

	zeroCnt := 0
	binaryCnt := 0

	for cur != "1" {
		// 0 제거
		zeroCnt += removeZeros()
		// 이진 변환
		binaryCnt += binaryChange()
	}

	return []int{binaryCnt, zeroCnt}
}

File: programmers/level2/python/이진변환반복하기.py
from collections import deque
from typing import List, Deque

cur: str = ""


def remove_zeros() -> int:
    global cur
    cnt: int = 0
    temp: List[str] = []

    for c in cur:
        if c == '0':
            cnt += 1
        else:
            temp.append(c)

    cur = ''.join(temp)
    return cnt


def binary_change() -> int:
    global cur
    length: int = len(cur)
    temp: Deque[str] = deque()

    while length > 0:
        temp.appendleft(str(length % 2))
        length = length // 2

    cur = ''.join(temp)
    return 1


def solution(s):
    global cur
    cur = s

    zero_cnt: int = 0
    binary_cnt: int = 0

    while cur != "1":
        zero_cnt += remove_zeros()
        binary_cnt += binary_change()

    return [binary_cnt, zero_cnt]